### PR TITLE
[HCFRO-274] Add progressbar

### PR DESCRIPTION
**Pull request by calinmiclaus**

### CODE REVIEW

**Comment by mook-as** [comment]
Hi!

Would it be reasonable to add tests for the new functionality to ensure we don't accidentally break it in the future?

And since I'm unfortunately lacking in context, I'll also be asking a few dumb questions inline.

File: schema/schema.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	ui terminal.UI
+}
+
+func NewSchemaParser(userInterface terminal.UI) *SchemaParser {
+	return &SchemaParser{
+		ui: userInterface,
+	}
+}
+
+func (p *SchemaParser) ParseSchema(schema string) (string, error) {
```
Comment (mook-as) [inline schema/schema.go]: Dumb question time!

Is this implementing an existing external interface? If not, would it make sense for the `string` argument / return value to be `[]byte`s instead?

File: schema/schema.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	if err != nil {
+		return "", err
+	}
+	return string(jsonResult), nil
+}
+
+func (p *SchemaParser) parseObject(key string, input map[string]interface{}) (map[string]interface{}, error) {
+	var result map[string]interface{} = make(map[string]interface{})
+
+	if prop, ok := input["properties"]; ok {
```
Comment (mook-as) [inline schema/schema.go]: Would the method be clearer if we early-returned if `!ok` instead?

File: schema/schema.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+		for k, v := range prop.(map[string]interface{}) {
+			t := v.(map[string]interface{})["type"]
```
Comment (mook-as) [inline schema/schema.go]: Is it alright to panic if `v` is not a `map[string]interface{}`? (I'm unclear on where the input would come from)

File: schema/schema.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+		return "", err
+	}
+	return string(jsonResult), nil
+}
+
+func (p *SchemaParser) parseObject(key string, input map[string]interface{}) (map[string]interface{}, error) {
+	var result map[string]interface{} = make(map[string]interface{})
+
+	if prop, ok := input["properties"]; ok {
+		for k, v := range prop.(map[string]interface{}) {
```
Comment (mook-as) [inline schema/schema.go]: Would the code be clearer if these variables have longer names (e.g. `key`, `value`)? (`type` is a keyword, though, unfortunately…)

Comment (mook-as) [inline schema/schema.go]: Oh, `key` is already a thing. Fun!

File: schema/schema.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+	}
+	jsonResult, err := json.Marshal(res)
+	if err != nil {
+		return "", err
+	}
+	return string(jsonResult), nil
+}
+
+func (p *SchemaParser) parseObject(key string, input map[string]interface{}) (map[string]interface{}, error) {
+	var result map[string]interface{} = make(map[string]interface{})
```
Comment (mook-as) [inline schema/schema.go]: Does `result := make(map[string]interface{})` not work?

File: schema/schema.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+				result[k] = nil
+			}
+		}
+	}
+
+	return result, nil
+}
+
+func isRequired(key string, input map[string]interface{}) bool {
+	required := input["required"].([]interface{})
```
Comment (mook-as) [inline schema/schema.go]: What should happen if `input["required"]` is missing or of the wrong type?

File: termprogressbar/progressbar.go [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+	"github.com/cheggaaa/pb"
+)
+
+type termuiprogressbar struct {
+	progressbar *pb.ProgressBar
+	Visible     bool
+}
+
+func NewProgressBar(total int, visible bool) *termuiprogressbar {
+	pbar := termuiprogressbar{}
```
Comment (mook-as) [inline termprogressbar/progressbar.go]: Wouldn't just returning a literal be cleaner?

``` golang
return &termuiprogressbar{
    progressbar: pb.New64(int64(total)),
    Visible: visible,
}
```

(Obviously I didn't go through the `make vet lint` cycle…)

File: termprogressbar/progressbar.go [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+func (pb *termuiprogressbar) Start() {
+	if pb.Visible {
+		pb.progressbar.Start()
+	}
+}
+
+func (pb *termuiprogressbar) Increment() {
+	if pb.Visible {
+		pb.progressbar.Increment()
```
Comment (mook-as) [inline termprogressbar/progressbar.go:30]: What happens if `Start()` was called while not visible, and `Increment()` was called when it is?

File: termprogressbar/progressbar.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+	}
+}
+
+func (pb *termuiprogressbar) FinishPrint(str string) {
+	if pb.Visible {
+		pb.progressbar.FinishPrint(str)
```
Comment (mook-as) [inline termprogressbar/progressbar.go:37]: What happens if `Start()` was called while not visible, and `FinishPrint()` called when it is?
What happens if `Start()` / `Increment()` was called while visible, and `FinishPrint()` called when it isn't?

**Comment by mook-as** [comment]
Overall comments:
- This looks like it wants to be two independent PRs
- It is unclear what the intended use cases are
  - What sort of input does `schema` want?
  - What is `termprogressbar` trying to do, that just using `github.com/cheggaaa/pb` directly wouldn't give?

File: termprogressbar/progressbar.go (hunk 9 above)
Comment (calinmiclaus) [inline termprogressbar/progressbar.go:37]: I made the visible property private, it didnt make sense to have it public

File: termprogressbar/progressbar.go (hunk 7 above)
Comment (calinmiclaus) [inline termprogressbar/progressbar.go]: made the change, looks cleaner now, thanks

**Comment by calinmiclaus** [comment]
I've split the progressbar and schema into 2 PRs

The only thing I added to github.com/cheggaaa/pb is a 'visible' property.

File: termprogressbar/progressbar_test.go [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+	"github.com/stretchr/testify/assert"
+)
+
+// tests if the progressbar is visible if it is initialized with visible=true
+func TestVisible(t *testing.T) {
+	assert := assert.New(t)
+
+	old := os.Stdout
+	r, w, _ := os.Pipe()
+	os.Stdout = w
```
Comment (mook-as) [inline termprogressbar/progressbar_test.go:21]: Consider doing `defer func() { os.Stdout = old }()` to ensure we never end up not restoring it?

**Comment by mook-as** [comment]
Thanks!
Still not convinced the `pb` wrapper should live in the `termui` package (as opposed to just having the user import `pb` directly — again, I can't tell what the use case is), but the code is almost ready to merge :)

File: termprogressbar/progressbar.go [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+import (
+	"github.com/cheggaaa/pb"
+)
+
+type termuiprogressbar struct {
+	progressbar *pb.ProgressBar
+	visible     bool
+}
+
+// creates a new progressbar
```
Comment (mook-as) [inline termprogressbar/progressbar.go]: `make lint` says:

> termui/termprogressbar/progressbar.go:12:1: comment on exported function NewProgressBar should be of the form "NewProgressBar ..."
> termui/termprogressbar/progressbar.go:13:46: exported func NewProgressBar returns unexported type *termuiprogressbar.termuiprogressbar, which can be annoying to use

**Comment by calinmiclaus** [comment]
Well, the only use case would be the user not wanting to have verbose output (--quiet flag passed somewhere). The vanilla pb package doesn't provide this functionality...

**Comment by mook-as** [comment]
Wouldn't that be [pb.NotPrint](https://godoc.org/github.com/cheggaaa/pb#ProgressBar)?

**Comment by calinmiclaus** [comment]
with NotPrint set to true, it does print out some info, although not with a progressbar. Its more convenient to have a wrapper around this than to fork pb and maintain the fork

File: termprogressbar/progressbar_test.go [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+	outC := make(chan string)
+
+	go func() {
+		var buf bytes.Buffer
+		io.Copy(&buf, r)
+		outC <- buf.String()
+	}()
+
+	w.Close()
+	defer func() { os.Stdout = old }()
```
Comment (mook-as) [inline termprogressbar/progressbar_test.go]: The `defer` should live next to the set (so if we panic right after for some reason it gets picked up).  The point is to let the reader not have to think about checking for it later, not to replace the `Close()` call at the same spot :)

Comment (calinmiclaus) [inline termprogressbar/progressbar_test.go]: I amended my last commit fixing this, thanks for the clear up :)

**Comment by mook-as** [comment]
Thanks for being patient with me as we go through the changes :)

Merged.

**Pull request merged**